File: app/src/main/java/info/nightscout/androidaps/plugins/NSClientInternal/NsClientReceiverDelegate.java
package info.nightscout.androidaps.plugins.NSClientInternal;

import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.wifi.WifiManager;

import com.squareup.otto.Bus;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.events.EventChargingState;
import info.nightscout.androidaps.events.EventNetworkChange;
import info.nightscout.androidaps.events.EventPreferenceChange;
import info.nightscout.androidaps.receivers.ChargingStateReceiver;
import info.nightscout.androidaps.receivers.NetworkChangeReceiver;
import info.nightscout.utils.SP;

class NsClientReceiverDelegate {

    private final Context context;
    private final Bus bus;

    private NetworkChangeReceiver networkChangeReceiver = new NetworkChangeReceiver();
    private ChargingStateReceiver chargingStateReceiver = new ChargingStateReceiver();

    private boolean allowedChargingState = true;
    private boolean allowedNetworkState = true;
    boolean allowed = true;

    NsClientReceiverDelegate(Context context, Bus bus) {
        this.context = context;
        this.bus = bus;
    }

    void registerReceivers() {
        Context context = MainApp.instance().getApplicationContext();
        // register NetworkChangeReceiver --> https://developer.android.com/training/monitoring-device-state/connectivity-monitoring.html
        // Nougat is not providing Connectivity-Action anymore ;-(
        context.registerReceiver(networkChangeReceiver,
                new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
        context.registerReceiver(networkChangeReceiver,
                new IntentFilter(WifiManager.WIFI_STATE_CHANGED_ACTION));

        EventNetworkChange event = networkChangeReceiver.grabNetworkStatus(context);
        if (event != null)
            bus.post(event);

        context.registerReceiver(chargingStateReceiver,
                new IntentFilter(Intent.ACTION_POWER_CONNECTED));
        context.registerReceiver(chargingStateReceiver,
                new IntentFilter(Intent.ACTION_POWER_DISCONNECTED));

        EventChargingState eventChargingState = chargingStateReceiver.grabChargingState(context);
        if (eventChargingState != null)
            bus.post(eventChargingState);

    }

    void unregisterReceivers() {
        context.unregisterReceiver(networkChangeReceiver);
        context.unregisterReceiver(chargingStateReceiver);
    }

    void onStatusEvent(EventPreferenceChange ev) {
        if (ev.isChanged(R.string.key_ns_wifionly) ||
                ev.isChanged(R.string.key_ns_wifi_ssids) ||
                ev.isChanged(R.string.key_ns_allowroaming)
                ) {
            EventNetworkChange event = networkChangeReceiver.grabNetworkStatus(MainApp.instance().getApplicationContext());
            if (event != null)
                bus.post(event);
        } else if (ev.isChanged(R.string.key_ns_chargingonly)) {
            EventChargingState event = chargingStateReceiver.grabChargingState(MainApp.instance().getApplicationContext());
            if (event != null)
                bus.post(event);
        }
    }

    void onStatusEvent(final EventChargingState ev) {
        boolean newChargingState = calculateStatus(ev);

        if (newChargingState != allowedChargingState) {
            allowedChargingState = newChargingState;
            processStateChange();
        }
    }

    void onStatusEvent(final EventNetworkChange ev) {
        boolean newNetworkState = calculateStatus(ev);

        if (newNetworkState != allowedNetworkState) {
            allowedNetworkState = newNetworkState;
            processStateChange();
        }
    }

    void processStateChange() {
        boolean newAllowedState =  allowedChargingState && allowedNetworkState;
        if (newAllowedState != allowed) {
            allowed = newAllowedState;
            bus.post(new EventPreferenceChange(R.string.key_nsclientinternal_paused));
        }
    }

    boolean calculateStatus(final EventChargingState ev) {
        boolean chargingOnly = SP.getBoolean(R.string.key_ns_chargingonly, false);

        boolean newAllowedState = true;

        if (!ev.isCharging && chargingOnly) newAllowedState = false;

        return newAllowedState;
    }

    boolean calculateStatus(final EventNetworkChange ev) {
        boolean wifiOnly = SP.getBoolean(R.string.key_ns_wifionly, false);
        String allowedSSIDs = SP.getString(R.string.key_ns_wifi_ssids, "");
        boolean allowRoaming = SP.getBoolean(R.string.key_ns_allowroaming, true);

        boolean newAllowedState = true;

        if (ev.wifiConnected) {
            if (!allowedSSIDs.trim().isEmpty() && !allowedSSIDs.contains(ev.ssid)) {
                newAllowedState = false;
            }
        } else {
            if ((!allowRoaming && ev.roaming) || wifiOnly) {
                newAllowedState = false;
            }
        }


        return newAllowedState;
    }

}

File: app/src/main/java/info/nightscout/androidaps/plugins/sync/nsclient/NsClientReceiverDelegate.kt
package info.nightscout.androidaps.plugins.sync.nsclient

import info.nightscout.androidaps.R
import info.nightscout.androidaps.events.EventChargingState
import info.nightscout.androidaps.events.EventNetworkChange
import info.nightscout.androidaps.events.EventPreferenceChange
import info.nightscout.androidaps.plugins.bus.RxBus
import info.nightscout.androidaps.receivers.ReceiverStatusStore
import info.nightscout.androidaps.interfaces.ResourceHelper
import info.nightscout.shared.sharedPreferences.SP
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class NsClientReceiverDelegate @Inject constructor(
    private val rxBus: RxBus,
    private val rh: ResourceHelper,
    private val sp: SP,
    private val receiverStatusStore: ReceiverStatusStore
) {

    private var allowedChargingState = true
    private var allowedNetworkState = true
    var allowed = true
    var blockingReason = ""

    fun grabReceiversState() {
        receiverStatusStore.updateNetworkStatus()
    }

    fun onStatusEvent(ev: EventPreferenceChange) {
        when {
            ev.isChanged(rh, R.string.key_ns_wifi) ||
                ev.isChanged(rh, R.string.key_ns_cellular) ||
                ev.isChanged(rh, R.string.key_ns_wifi_ssids) ||
                ev.isChanged(rh, R.string.key_ns_allow_roaming) -> {
                receiverStatusStore.updateNetworkStatus()
                receiverStatusStore.lastNetworkEvent?.let { onStatusEvent(it) }
            }

            ev.isChanged(rh, R.string.key_ns_charging) ||
                ev.isChanged(rh, R.string.key_ns_battery)       -> {
                receiverStatusStore.broadcastChargingState()
            }
        }
    }

    fun onStatusEvent(ev: EventChargingState) {
        val newChargingState = calculateStatus(ev)
        if (newChargingState != allowedChargingState) {
            allowedChargingState = newChargingState
            blockingReason = rh.gs(R.string.blocked_by_charging)
            processStateChange()
        }
    }

    fun onStatusEvent(ev: EventNetworkChange) {
        val newNetworkState = calculateStatus(ev)
        if (newNetworkState != allowedNetworkState) {
            allowedNetworkState = newNetworkState
            blockingReason = rh.gs(R.string.blocked_by_connectivity)
            processStateChange()
        }
    }

    private fun processStateChange() {
        val newAllowedState = allowedChargingState && allowedNetworkState
        if (newAllowedState != allowed) {
            allowed = newAllowedState
            rxBus.send(EventPreferenceChange(rh.gs(R.string.key_nsclientinternal_paused)))
        }
    }

    fun calculateStatus(ev: EventChargingState): Boolean =
        !ev.isCharging && sp.getBoolean(R.string.key_ns_battery, true) ||
            ev.isCharging && sp.getBoolean(R.string.key_ns_charging, true)

    fun calculateStatus(ev: EventNetworkChange): Boolean =
        ev.mobileConnected && sp.getBoolean(R.string.key_ns_cellular, true) && !ev.roaming ||
            ev.mobileConnected && sp.getBoolean(R.string.key_ns_cellular, true) && ev.roaming && sp.getBoolean(R.string.key_ns_allow_roaming, true) ||
            ev.wifiConnected && sp.getBoolean(R.string.key_ns_wifi, true) && sp.getString(R.string.key_ns_wifi_ssids, "").isEmpty() ||
            ev.wifiConnected && sp.getBoolean(R.string.key_ns_wifi, true) && sp.getString(R.string.key_ns_wifi_ssids, "").split(";").contains(ev.ssid)
}